src/explorer.api/Explorers/ColumnExplorer.cs

```diff
@@ -2,71 +2,48 @@ namespace Explorer
 {
     using System;
     using System.Collections.Generic;
-    using System.Collections.Concurrent;
     using System.Threading.Tasks;
 
-    using Aircloak.JsonApi;
-    using Aircloak.JsonApi.ResponseTypes;
-    using Explorer.Api.Models;
-
-    internal abstract class ColumnExplorer
+    internal class ColumnExplorer
     {
-        private readonly ConcurrentStack<ExploreResult> exploreResults;
+        private readonly Dictionary<Type, Task> childTasks;
+
+        private readonly Dictionary<Type, ExplorerImpl> childExplorers;
 
-        protected ColumnExplorer(JsonApiClient apiClient, ExploreParams exploreParams)
+        public ColumnExplorer()
         {
-            ApiClient = apiClient;
-            ExploreParams = exploreParams;
             ExplorationGuid = Guid.NewGuid();
 
-            exploreResults = new ConcurrentStack<ExploreResult>();
-            exploreResults.Push(new ExploreResult(ExplorationGuid, Status.New));
+            childTasks = new Dictionary<Type, Task>();
+            childExplorers = new Dictionary<Type, ExplorerImpl>();
         }
 
         public Guid ExplorationGuid { get; }
 
-        public ExploreResult LatestResult
+        public IEnumerable<IExploreMetric> ExploreMetrics
         {
             get
             {
-                if (exploreResults.TryPeek(out var result))
+                foreach (var explorer in childExplorers.Values)
                 {
-                    return result;
+                    foreach (var metric in explorer.Metrics)
+                    {
+                        yield return metric;
+                    }
                 }
-                else
-                {
-                    return new ExploreError(ExplorationGuid, "Unexpected error: unknown explorer status.");
-                }
-            }
-
-            protected set
-            {
-                exploreResults.Push(value);
             }
         }
 
-        protected ExploreParams ExploreParams { get; }
-
-        protected JsonApiClient ApiClient { get; }
-
-        public abstract Task Explore();
-
-        protected async Task<QueryResult<TResult>> ResolveQuery<TResult>(
-            IQuerySpec<TResult> query,
-            TimeSpan timeout)
+        public Task Completion()
```

Comment:
Here, I would call the `Task.WhenAll(childTasks.Values)` in the constructor and keep the result inside a member of the class. Then I would make this `Completion()` method into a property that returns just the task status.

Reply:
Yes I think this is clearer, I'll do it.

Reply:
Commit 09820d92584840a082f2236cd805a571997a4c1f